orttraining/orttraining/core/framework/gradient_graph_builder.cc

```diff
@@ -255,21 +256,84 @@ Status GradientGraphBuilder::Build(const std::unordered_set<std::string>* p_init
 
   // so far, visited are the minimum node in between
   // visited_node_args are the node_args involved
+  std::unordered_map<std::string, int> input_arg_ref_count;
+  for (auto node : visited) {
```

Comment:
refactor to have a two level loop? 

for (pass : [0,1]) :
   for (node : visited) :
       if (pass != pass_table[node])
          continue;